### PR TITLE
[stdlib][string] Add benchmark for String(repeating:count:)

### PR DESCRIPTION
**Pull request by valeriyvan**
Adds benchmark for String(repeating:count:)

### CODE REVIEW

**Comment by valeriyvan** [comment]
Should help benchmark #32369

**Comment by CodaFi** [comment]
@swift-ci smoke benchmark

**Comment by xwu** [comment]
...I thought the new benchmark should appear here in the summarized report?

**Comment by CodaFi** [comment]
It's possible it's not in the smoke suite. Let's try

@swift-ci benchmark

**Comment by valeriyvan** [comment]
ping

**Comment by natecook1000** [comment]
@gottesmm Do you see any reason why this doesn’t seem to be showing up as “Added” in the benchmark run?

**Comment by gottesmm** [comment]
@natecook1000 Not sure. Is it running at all?

@valeriyvan here is my suggestion. Can you try building this locally? 

From the README (with my additions):

```
1. `$ mkdir build; cd build`                                                                                         
2. `$ cmake [path to swift src]/benchmark -G Ninja -DSWIFT_EXEC=[path to built swiftc]`                              
3. `$ ninja swift-benchmark-macosx-x86_64
4. `$ ./bin/Benchmark_O-* --list | grep StringRepeating
```

**Comment by gottesmm** [comment]
(I am suggesting that you do it separately just to save you build time)

File: benchmark/single-source/StringRepeating.swift [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+        blackHole(repeating("x", count: 10))
+    }
+}
+
+@inline(never)
+public func run_singleAsciiCharacterCount1(N: Int) {
+    for _ in 1...5000*N {
+        blackHole(repeating("x", count: 1))
+    }
+}
```
Comment (milseman) [inline benchmark/single-source/StringRepeating.swift:66]: Are you seeing any meaningful difference between this and the 10 versions above? That would be surprising.

File: benchmark/single-source/StringRepeating.swift [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+  BenchmarkInfo(name: "StringRepeating/U1F1F8U1F1FACount2",
+                runFunction: run_U1F1F8U1F1FACount2,
+                tags: [.validation, .api, .String]),
+  BenchmarkInfo(name: "StringRepeating/U301cafeCount5",
+                runFunction: run_U301cafeCount5,
+                tags: [.validation, .api, .String]),
+  BenchmarkInfo(name: "StringRepeating/LongMixedStringCount100",
+                runFunction: run_longMixedStringCount100,
+                tags: [.validation, .api, .String])
+]
```
Comment (milseman) [inline benchmark/single-source/StringRepeating.swift:46]: What is the rationale for this set of cases? Some seem redundant, and I would think we'd be more interested in small vs non-small forms

**Comment by valeriyvan** [comment]
> @natecook1000 Not sure. Is it running at all?
> 
> @valeriyvan here is my suggestion. Can you try building this locally?
> 
> From the README (with my additions):
> 
> ```
> 1. `$ mkdir build; cd build`                                                                                         
> 2. `$ cmake [path to swift src]/benchmark -G Ninja -DSWIFT_EXEC=[path to built swiftc]`                              
> 3. `$ ninja swift-benchmark-macosx-x86_64
> 4. `$ ./bin/Benchmark_O-* --list | grep StringRepeating
> ```

I've failed run benchmarks built with these steps. Something wrong with path to `libswiftCore.dylib`.
When I run with `./swift/utils/build-script -R -B` I see this at end:

```
888 StringMatch                                   55   46600  46700      46800  48100   54800     245760
889 StringRemoveDupes                             58     809    814        835    870    1031     167936
Traceback (most recent call last):
  File "/Users/valeriyvan/Documents/Apple/build/Ninja-ReleaseAssert/swift-macosx-x86_64/bin/Benchmark_Driver", line 970, in <module>
    exit(main())
  File "/Users/valeriyvan/Documents/Apple/build/Ninja-ReleaseAssert/swift-macosx-x86_64/bin/Benchmark_Driver", line 966, in main
    return args.func(args)
  File "/Users/valeriyvan/Documents/Apple/build/Ninja-ReleaseAssert/swift-macosx-x86_64/bin/Benchmark_Driver", line 299, in run_benchmarks
    csv_log = driver.run_and_log(csv_console=(args.output_dir is None))
  File "/Users/valeriyvan/Documents/Apple/build/Ninja-ReleaseAssert/swift-macosx-x86_64/bin/Benchmark_Driver", line 286, in run_and_log
    result = result_values(self.run_independent_samples(test))
  File "/Users/valeriyvan/Documents/Apple/build/Ninja-ReleaseAssert/swift-macosx-x86_64/bin/Benchmark_Driver", line 217, in run_independent_samples
    for _ in range(self.args.independent_samples)
  File "/Users/valeriyvan/Documents/Apple/build/Ninja-ReleaseAssert/swift-macosx-x86_64/bin/Benchmark_Driver", line 171, in run
    return list(results.items())[0][1] if test else results
IndexError: list index out of range
ninja: build stopped: subcommand failed.
Building the standard library for: swift-test-stdlib-macosx-x86_64
Running Swift benchmarks for: check-swift-benchmark-macosx-x86_64
ERROR: command terminated with a non-zero exit status 1, aborting
```

Something wrong with benchmark but I can't figure out exactly what.

**Comment by shahmishal** [comment]
Please update the base branch to `main` by Oct 5th otherwise the pull request will be closed automatically. 

* How to change the base branch: ([Link](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/changing-the-base-branch-of-a-pull-request))
* More detail about the branch update: ([Link](https://forums.swift.org/t/updating-branch-names/40412/))

File: benchmark/single-source/StringRepeating.swift [hunk 3]
```diff
@@ -0,0 +1,143 @@
+//===--- StringRepeating.swift -------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2019 Apple Inc. and the Swift project authors
```
Comment (xwu) [inline benchmark/single-source/StringRepeating.swift:5]:
```suggestion
// Copyright (c) 2014 - 2023 Apple Inc. and the Swift project authors
```

File: benchmark/single-source/StringRepeating.swift (hunk 1 above)
Comment (valeriyvan) [reply benchmark/single-source/StringRepeating.swift:66]: Agree. This edge case would be interesting for testing, but not for benchmarking.
Removed.

File: benchmark/single-source/StringRepeating.swift (hunk 2 above)
Comment (valeriyvan) [reply benchmark/single-source/StringRepeating.swift:28]: reduced number of benchmarks to 4

**Comment by valeriyvan** [comment]
ping

File: benchmark/single-source/StringRepeating.swift [hunk 4]
```diff
@@ -0,0 +1,83 @@
+//===--- StringRepeating.swift -------------------------------------------===//
```
Comment (xwu) [inline benchmark/single-source/StringRepeating.swift:1]:
```suggestion
//===--- StringRepeating.swift --------------------------------------------===//
```

File: benchmark/single-source/StringRepeating.swift [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+                runFunction: run_33CyrillicCharactersCount2,
+                tags: [.validation, .api, .String]),
+  BenchmarkInfo(name: "String.initRepeating.longMixedString.Count100",
+                runFunction: run_longMixedStringCount100,
+                tags: [.validation, .api, .String])
+]
+
+@inline(never)
+func repeating(_ i: String, count: Int) -> String {
+  let s = String(repeating: getString(i), count: count)
```
Comment (xwu) [inline benchmark/single-source/StringRepeating.swift:32]: What is the purpose of this function? Why must it never be inlined?

File: benchmark/single-source/StringRepeating.swift [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+@inline(never)
+public func run_33CyrillicCharactersCount2(N: Int) {
+    for _ in 1...5000*N {
+        blackHole(repeating("абвгґдеєжзиіїйклмнопрстуфхцчшщьюя", count: 2))
+    }
+}
+
+func getLongString() -> String {
+  let long = """
+    Swift is a multi-paradigm, compiled programming language created for
```
Comment (xwu) [inline benchmark/single-source/StringRepeating.swift:59]: This looks like copyrighted text. Did you secure permission to re-license this here? If not, please find public domain multilingual text. I _believe_ the Universal Declaration of Human Rights is one example but you would need to double-check.

File: benchmark/single-source/StringRepeating.swift [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+     which allows C, Objective-C, C++ and Swift code to run within a single
+     program.
+     Існує багато варіацій уривків з Lorem Ipsum, але більшість з них зазнала
+     певних змін на кшталт жартівливих вставок або змішування слів, які навіть
+     не виглядають правдоподібно.
+     日本語の場合はランダムに生成された文章以外に、
+     著作権が切れた小説などが利用されることもある。
+     🦩
+    """
+  return getString(long)
```
Comment (xwu) [inline benchmark/single-source/StringRepeating.swift:75]: Does this function really need to be factored out of the single call site? For what purpose?

**Review by xwu — COMMENTED** [review]
See comments inline. Not sure who the code owner is today but @Catfish-Man might be able to direct us.

File: benchmark/single-source/StringRepeating.swift (hunk 6 above)
Comment (valeriyvan) [reply benchmark/single-source/StringRepeating.swift:59]: This exact string appears in following files in this repo:
* `benchmark/single-source/CString.swift`
* `benchmark/single-source/UTF8Decode.swift`
* `benchmark/single-source/StringBuilder.swift`
* `benchmark/single-source/StringSplitting.swift`
* `validation-test/compiler_crashers_2_fixed/0109-issue-47314.swift`
* `test/Prototypes/UnicodeDecoders.swift`
This suggests it should be safe to use this string once again.

File: benchmark/single-source/StringRepeating.swift (hunk 4 above)
Comment (valeriyvan) [reply benchmark/single-source/StringRepeating.swift]: fixed

File: benchmark/single-source/StringRepeating.swift (hunk 5 above)
Comment (valeriyvan) [reply benchmark/single-source/StringRepeating.swift:32]: Many, if not all benchmarks for `String` follow the sma pattern. Take, e.g. `benchmark/single-source/StringBuilder.swift`. Here I just follow pattern.
As I understand, `@inline(never)` is needed here to forbid some optimizations.

File: benchmark/single-source/StringRepeating.swift (hunk 7 above)
Comment (valeriyvan) [reply benchmark/single-source/StringRepeating.swift]: fixed

File: benchmark/single-source/StringRepeating.swift (hunk 5 above)
Comment (xwu) [reply benchmark/single-source/StringRepeating.swift:32]: I'm not seeing any examples in the file you've referred to where a `String` API is wrapped in a function like this. Can you link to the exact line you're thinking of?

File: benchmark/single-source/StringRepeating.swift [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+                runFunction: run_longMixedStringCount100,
+                tags: [.validation, .api, .String])
+]
+
+@inline(never)
+public func run_singleAsciiCharacterCount100(N: Int) {
+  let string = getString("x")
+  for _ in 1...5000*N {
+    blackHole(String(repeating: string, count: 100))
+  }
```
Comment (xwu) [inline benchmark/single-source/StringRepeating.swift:35]:
```suggestion
  let string = "x"
  for _ in 1...5000*N {
    blackHole(String(repeating: getString(string), count: 100))
  }
```

File: benchmark/single-source/StringRepeating.swift [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+}
+
+@inline(never)
+public func run_26AsciiCharactersCount2(N: Int) {
+  let string = getString("abcdefghijklmnopqrstuvwxyz")
+  for _ in 1...5000*N {
+    blackHole(String(repeating: string, count: 2))
+  }
```
Comment (xwu) [inline benchmark/single-source/StringRepeating.swift:43]:
```suggestion
  let string = "abcdefghijklmnopqrstuvwxyz"
  for _ in 1...5000*N {
    blackHole(String(repeating: getString(string), count: 2))
  }
```

File: benchmark/single-source/StringRepeating.swift [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+}
+
+@inline(never)
+public func run_33CyrillicCharactersCount2(N: Int) {
+  let string = getString("абвгґдеєжзиіїйклмнопрстуфхцчшщьюя")
+  for _ in 1...5000*N {
+    blackHole(String(repeating: string, count: 2))
+  }
```
Comment (xwu) [inline benchmark/single-source/StringRepeating.swift:51]:
```suggestion
  let string = "абвгґдеєжзиіїйклмнопрстуфхцчшщьюя"
  for _ in 1...5000*N {
    blackHole(String(repeating: getString(string), count: 2))
  }
```

File: benchmark/single-source/StringRepeating.swift [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+    program.
+    Існує багато варіацій уривків з Lorem Ipsum, але більшість з них зазнала
+    певних змін на кшталт жартівливих вставок або змішування слів, які навіть
+    не виглядають правдоподібно.
+    日本語の場合はランダムに生成された文章以外に、
+    著作権が切れた小説などが利用されることもある。
+    🦩
+    """
+  for _ in 1...5000*N {
+    blackHole(String(repeating: string, count: 100))
```
Comment (xwu) [inline benchmark/single-source/StringRepeating.swift:74]:
```suggestion
    blackHole(String(repeating: getString(string), count: 100))
```

**Review by xwu — CHANGES_REQUESTED** [review]
You mustn't hoist `getString` out of the loop: if you do, then the optimizer will be able to recognize that you're always passing `blackHole` the same argument, meaning that it can then hoist `String(repeating:count:)` out of the loop as well and defeat your benchmark.

File: benchmark/single-source/StringRepeating.swift (hunk 8 above)
Comment (valeriyvan) [reply benchmark/single-source/StringRepeating.swift]: fixed

File: benchmark/single-source/StringRepeating.swift (hunk 9 above)
Comment (valeriyvan) [reply benchmark/single-source/StringRepeating.swift]: fixed

File: benchmark/single-source/StringRepeating.swift (hunk 10 above)
Comment (valeriyvan) [reply benchmark/single-source/StringRepeating.swift]: fixed

File: benchmark/single-source/StringRepeating.swift (hunk 11 above)
Comment (valeriyvan) [reply benchmark/single-source/StringRepeating.swift]: fixed

**Comment by xwu** [comment]
@swift-ci please benchmark

**Comment by xwu** [comment]
@eeckstein Are you by chance the code owner now?

**Comment by xwu** [comment]
```
------- Performance (x86_64): -O -------

REGRESSION                                      OLD       NEW       DELTA                RATIO    
ArrayAppendGenericStructs                       593.333   740.0     +24.7%               **0.80x (?)**

IMPROVEMENT                                     OLD       NEW       DELTA                RATIO    
Data.init.Sequence.511B.Count.I                 32.094    29.156    -9.2%                **1.10x (?)**
ObjectiveCBridgeStubToNSStringRef               95.0      87.429    -8.0%                **1.09x (?)**
Data.init.Sequence.513B.Count.I                 32.29     29.923    -7.3%                **1.08x (?)**
Diffing.Myers.Similar                           367.25    341.167   -7.1%                **1.08x (?)**

ADDED                                           MIN       MAX       MEAN                 MAX_RSS  
String.initRepeating.1AsciiChar.Count100        13180.0   13382.0   13249.333333333334   —        
String.initRepeating.26AsciiChar.Count2         708.667   724.667   714.778              —        
String.initRepeating.33CyrillicChar.Count2      1453.0    1567.0    1496.6666666666667   —        
String.initRepeating.longMixedString.Count100   22220.0   28435.0   24516.0              —        

------- Code size: -O -------

------- Performance (x86_64): -Osize -------

REGRESSION                                      OLD       NEW       DELTA                RATIO    
FlattenListFlatMap                              2502.0    2823.0    +12.8%               **0.89x (?)**

IMPROVEMENT                                     OLD       NEW       DELTA                RATIO    
Dictionary4                                     202.0     160.111   -20.7%               **1.26x (?)**
Dictionary4OfObjects                            243.857   207.25    -15.0%               **1.18x (?)**
ObjectiveCBridgeStubDateAccess                  152.538   130.75    -14.3%               **1.17x (?)**
StringFromLongWholeSubstring                    2.935     2.719     -7.4%                **1.08x (?)**
ObjectiveCBridgeStubToNSStringRef               95.0      88.731    -6.6%                **1.07x (?)**

ADDED                                           MIN       MAX       MEAN                 MAX_RSS  
String.initRepeating.1AsciiChar.Count100        13565.0   14013.0   13811.333333333334   —        
String.initRepeating.26AsciiChar.Count2         663.333   666.667   664.6666666666666    —        
String.initRepeating.33CyrillicChar.Count2      833.0     848.667   841.778              —        
String.initRepeating.longMixedString.Count100   19761.0   19894.0   19838.333333333332   —        

------- Code size: -Osize -------

------- Performance (x86_64): -Onone -------

IMPROVEMENT                                     OLD       NEW       DELTA                RATIO    
NSError                                         391.667   366.0     -6.6%                **1.07x (?)**

ADDED                                           MIN       MAX       MEAN                 MAX_RSS  
String.initRepeating.1AsciiChar.Count100        13638.0   13747.0   13710.666666666666   —        
String.initRepeating.26AsciiChar.Count2         683.0     697.667   688.5556666666666    —        
String.initRepeating.33CyrillicChar.Count2      853.333   865.0     857.4443333333334    —        
String.initRepeating.longMixedString.Count100   17324.0   19086.0   18241.666666666668   —        

------- Code size: -swiftlibs -------


ERROR naming: 'String.initRepeating.longMixedString.Count100' name is 45 characters long.
INFO naming: Benchmark name should not be longer than 40 characters.
ERROR runtime: 'String.initRepeating.longMixedString.Count100' execution took at least 17193 μs.
INFO runtime: Decrease the workload of 'String.initRepeating.longMixedString.Count100' by a factor of 32 (100), to be less than 1000 μs.
INFO memory: unable to compute memory footprint of 'String.initRepeating.longMixedString.Count100'
INFO memory: unable to compute memory footprint of 'String.initRepeating.26AsciiChar.Count2'
ERROR naming: 'String.initRepeating.33CyrillicChar.Count2' name is 42 characters long.
INFO naming: Benchmark name should not be longer than 40 characters.
INFO memory: unable to compute memory footprint of 'String.initRepeating.33CyrillicChar.Count2'
ERROR runtime: 'String.initRepeating.1AsciiChar.Count100' execution took at least 12379 μs.
INFO runtime: Decrease the workload of 'String.initRepeating.1AsciiChar.Count100' by a factor of 16 (100), to be less than 1000 μs.
INFO memory: unable to compute memory footprint of 'String.initRepeating.1AsciiChar.Count100'

```

**Comment by xwu** [comment]
17 seconds _is_ pretty long for a single benchmark test. Can you decrease those by the suggested factors (so, something like 200 to 400 loop iterations instead of 5000)?

**Comment by xwu** [comment]
17 seconds _is_ pretty long for a single benchmark test. Can you decrease those by the suggested factors (so, something like 200 to 400 loop iterations instead of 5000)?

**Comment by eeckstein** [comment]
@swift-ci benchmark

**Comment by eeckstein** [comment]
@swift-ci smoke test

**Pull request merged**